### PR TITLE
feat(cache): Repository cache fingerprinting

### LINKED ISSUE #17247: Unrecoverable repo cache problem when repository reset
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Renovate gets into an unrecoverable problem when the following happens:
- Renovate runs on a repo, saves repo cache
- Repo is deleted then recreated (e.g. PRs start again at 1)

I think the problem lies in the PR cache, but this should be detectable because we fetch a page of PRs every run and should be able to recognize the mismatch and drop the existing cache.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- New cache repo revision
- New `fingerprint` parameter
- Gradual migration

## Context

- Closes: #17247

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Blocked by #17364

**Review by rarkins — COMMENTED** [review]
Why the need for gradual migration instead of full? Assumption that it's too heavy for the app in production?

**Comment by zharinov** [comment]
> Why the need for gradual migration instead of full? Assumption that it's too heavy for the app in production?

Yes, otherwise it will reset all the caches instantly 🤷‍♂️

**Review by viceice — COMMENTED** [review]
🤔 is writing new json files so bad? i don't see a big issue for the hoseted app. I think it's always copying the cache file after each repo anyways. 🤔

**Comment by zharinov** [comment]
It's not about writing JSON, we have to reset all the cache migrating from 12 to 13.

**Comment by viceice** [comment]
> It's not about writing JSON, we have to reset all the cache while migrating from 12 to 13. I forget to add this "resetting" part.

but didn't we still use / migrate the old cache? so it's irrelevant if we change it afterwards? 
or do you fully discard the old cache?

**Comment by zharinov** [comment]
If we migrate gradually, we have to migrate previous versions to 12, or reset the data and migrate it to 13 for the smaller fraction of repos. If we're okay to reset them all at once, I think we can make "fresh start" for the revisions.

**Comment by viceice** [comment]
i still don't know why we can't migrate all old caches like we did before?

btw the hosted app is using multiple worker nodes and every node will only process one repo at a time.

File: lib/util/cache/repository/index.spec.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { mocked } from '../../../../test/util';
 import { GlobalConfig } from '../../../config/global';
-import type { RenovateConfig } from '../../../config/types';
+import type { WorkerPlatformConfig } from '../../../workers/repository/init/apis';
```
Comment (viceice) [inline lib/util/cache/repository/index.spec.ts:3]: this is a bad idea, we don't like to import types from worker to utils.

we should define what we need here instead

File: lib/util/cache/repository/init.ts [hunk 2]
```diff
@@ -1,30 +1,37 @@
-import type { RenovateConfig } from '../../../config/types';
+import type { WorkerPlatformConfig } from '../../../workers/repository/init/apis';
```
Comment (viceice) [inline lib/util/cache/repository/init.ts:1]: same as above

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/cache/repository/common.ts [hunk 3]
```diff
@@ -1,50 +1,16 @@
...
-import type {
-  RepoCacheRecordV10,
-  RepoCacheRecordV11,
-  RepoCacheRecordV12,
-} from './types';
+import type { RepoCacheRecord } from './types';
 
 // Increment this whenever there could be incompatibilities between old and new cache structure
-export const CACHE_REVISION = 12;
+export const CACHE_REVISION = 1;
```
Comment (rarkins) [inline lib/util/cache/repository/common.ts:5]: Why not just keep incrementing? I know you can differentiate, but I'm not sure I like it

File: lib/workers/repository/init/index.ts [hunk 4]
```diff
@@ -32,9 +32,12 @@ export async function initRepo(
 ): Promise<RenovateConfig> {
   PackageFiles.clear();
   let config: RenovateConfig = initializeConfig(config_);
+
   await resetCaches();
-  config = await initApis(config);
-  await initializeCaches(config);
+  const workerPlatformConfig = await initApis(config);
+  config = workerPlatformConfig;
```
Comment (rarkins) [inline lib/workers/repository/init/index.ts:38]: Why these changes?

Comment (zharinov) [reply lib/workers/repository/init/index.ts:38]: `RenovateConfig` interface doesn't contain required fields, but in fact they're present after `initApis` call.

Comment (rarkins) [reply lib/workers/repository/init/index.ts:38]: I don't understand why the intermediate variable is needed though

Comment (zharinov) [reply lib/workers/repository/init/index.ts]: I wanted to avoid type assertion

**Review by rarkins — APPROVED** [review]
Hard to tell from code inspection alone

**Review by viceice — DISMISSED** [review]
i still don't understand why we can't migrate old cache this time

**Comment by zharinov** [comment]
Because for some repos, we may reuse the cache of deleted repository with same name, and there is no easy way to detect repos like these.

**Comment by viceice** [comment]
> Because for some repos, we may reuse the cache of previously deleted repository with same name, and there is no easy way to detect repos like these.

that case  is very rare, so those could be manually clean like today?

**Comment by rarkins** [comment]
How much work would it be to preserve existing caches? i.e. knowing that maybe some of them are invalid, but thinking it's an OK trade-off?

**Comment by zharinov** [comment]
Well, we may just write and check fingerprints only for repos that don't have any cache (i.e. only for new ones).

**Comment by rarkins** [comment]
So the choice is between:
- (current) burn the existing cache, but do it gradually to prevent overload, or
- preserve the existing cache, do it immediately: add the fingerprint to all existing caches?

**Comment by zharinov** [comment]
Not exactly:
- (current) reset all the cache at once and start with newest revision
- (previous) gradually migrate all the caches controlling pace of change via hardcoded variable
- add fingerprint to caches, despite some of them can be wrong

**Comment by viceice** [comment]
i would simply add the fingerprint if missing as the wrong cases are only a few repos.

**Review by rarkins — CHANGES_REQUESTED** [review]
OK, let's go with @viceice's suggestion and I'll keep manually fixing any we find in the app. Sorry for the changing of requirements

**Comment by zharinov** [comment]
No problem, I'll restore it

File: lib/util/cache/repository/impl/base.ts [hunk 5]
```diff
@@ -41,26 +56,30 @@ export abstract class RepoCacheBase implements RepoCache {
...
+      return;
+    }
+    await this.restoreFromRev12(oldCache);
   }
 
   async load(): Promise<void> {
     try {
-      const data = await this.read();
-      if (!is.string(data)) {
+      const rawOldCache = await this.read();
```
Comment (viceice) [inline lib/util/cache/repository/impl/base.ts:69]: any reason for this rename?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/util/cache/repository/impl/base.ts (hunk 5 above)
Comment (zharinov) [reply lib/util/cache/repository/impl/base.ts:69]: In order to avoid confusion with `this.data`

**Review by rarkins — APPROVED** [review]
